File: core0/us/sdk/wallet/cs/us/gov/crypto/ec.cs
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the AGPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
using Secp256k1Net;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Numerics;
using System.Runtime.InteropServices;
using System.Text;

namespace us.gov.crypto
{
    public class KeyPair
    {        
        public byte[] PrivateKey { get; set; }
        public byte[] PublicKey { get;set; }

        public KeyPair()
        {

        }

        public KeyPair(byte[] _public, byte[] _private)
        {
            this.PublicKey = _public;
            this.PrivateKey = _private;
        }
    }

    public class ec
    {        
        private readonly static ec _instance = new ec();

        private ec()
        {
        }

        public static ec Instance
        {
            get
            {
                return _instance;
            }
        }

        /*** Error PInvokeStackImbalance -->
           1. CTRL + ALT + E
           2. Bajo "Asistentes de depuración administrados" desmarque PInvokeStackImbalance.
        ***/

        public byte[] getPublicKey(byte[] priv)
        {
            byte[] ret = null;
            try
            {
                using (var secp256k1 = new Secp256k1())
                {
                    byte[] publicKey = new byte[64];
                    secp256k1.PublicKeyCreate(publicKey, priv);
                    byte[] publicKeyCompressed = new byte[33];
                    secp256k1.PublicKeySerialize(publicKeyCompressed, publicKey, Flags.SECP256K1_EC_COMPRESSED);
                    return publicKeyCompressed;
                }
            }
            catch (Exception)
            {     
                
            }
            return ret;
        }


        public KeyPair getKeyPair()
        {
           KeyPair ret = new KeyPair();
            try
            {
                using (var secp256k1 = new Secp256k1())
                {
                    byte[] privateKey = new byte[32];
                    var rnd = System.Security.Cryptography.RandomNumberGenerator.Create();
                    do { rnd.GetBytes(privateKey); }
                    while (!secp256k1.SecretKeyVerify(privateKey));
                    byte[] publicKey = new byte[64];
                    secp256k1.PublicKeyCreate(publicKey, privateKey);
                    byte[] publicKeyCompressed = new byte[33];
                    secp256k1.PublicKeySerialize(publicKeyCompressed, publicKey, Flags.SECP256K1_EC_COMPRESSED);
                    ret = new KeyPair(publicKeyCompressed, privateKey);
                }
            }
            catch (Exception)
            {

            }
            return ret;
        }

        public byte[] generate_shared_key(byte[] priv, byte[] pub, int length) {
            try
            {
                using (var secp256k1 = new Secp256k1())
                {
                    byte[] result = new Byte[32];
                    if (secp256k1.Ecdh(result, pub, priv))
                    {
                        return result;
                    }
                }
            }
            catch (Exception)
            {             
            }
            return null;
        }

        public String to_b58(byte[] k)
        {
            try
            {
                return base58.encode(k);
            }
            catch (Exception)
            {
                return null;
            }
        }

        public bool verify(byte[] pub, String text, String signature_der_b58)
        {
            try
            {
                using (var secp256k1 = new Secp256k1())
                {
                    byte[] signatureOutput = new byte[Secp256k1.SIGNATURE_LENGTH];
                    byte[] pubkeyParsed = new byte[Secp256k1.PUBKEY_LENGTH];
                    var messageBytes = Encoding.UTF8.GetBytes(text);
                    var messageHash = System.Security.Cryptography.SHA256.Create().ComputeHash(messageBytes);
                    //secp256k1.SignatureNormalize()
                    var DERsignature = base58.decode(signature_der_b58);
                    Debug.Assert(secp256k1.PublicKeyParse(pubkeyParsed, pub));
                    Debug.Assert(secp256k1.SignatureParseDer(signatureOutput, DERsignature));
                    return secp256k1.Verify(signatureOutput, messageHash, pubkeyParsed);
                }
            }
            catch (Exception e)
            {
                System.Diagnostics.Debug.WriteLine(e.Message);
                return false;
            }
        }

        //delegate int secp256k1_ecdsa_signature_serialize_der(const secp256k1_context* ctx, unsigned char* output, size_t *outputlen, const secp256k1_ecdsa_signature* sig) 

        public byte[] sign(byte[] priv, String message)
        {
            try
            {
                using (var secp256k1 = new Secp256k1())
                {
                    var messageBytes = Encoding.UTF8.GetBytes(message);
                    var messageHash = System.Security.Cryptography.SHA256.Create().ComputeHash(messageBytes);
                    var signature = new byte[64];
                    secp256k1.Sign(signature, messageHash, priv);

                    byte[] signatureDer = new byte[Secp256k1.SERIALIZED_DER_SIGNATURE_MAX_SIZE];
                    int outL = 0;
                    Debug.Assert(secp256k1.SignatureSerializeDer(signatureDer, signature, out outL));
                    Array.Resize(ref signatureDer, outL);

                    return signatureDer; 
                }
            }
            catch (Exception)
            {
                //System.out.println(e.getMessage());
                return new byte[0];
            }
        }

        public String sign_encode(byte[] priv, String message)
        {
            return base58.encode(sign(priv, message));
        }
    }

    public static class Extensions
    {
        public static string ToHexString(this byte[] span)
        {
            return "0x" + BitConverter.ToString(span).Replace("-", "").ToLowerInvariant();
        }

        public static byte[] HexToBytes(this string hexString)
        {
            int chars = hexString.Length;
            byte[] bytes = new byte[chars / 2];
            for (int i = 0; i < chars; i += 2)
            {
                bytes[i / 2] = Convert.ToByte(hexString.Substring(i, 2), 16);
            }
            return bytes;
        }
    }

    public abstract class BigIntegerConverter
    {
        /// <summary>
        /// Obtains the bytes that represent the BigInteger as if it was a big endian 256-bit integer.
        /// </summary>
        /// <param name="bigInteger">The BigInteger to obtain the byte representation of.</param>
        /// <returns>Returns the bytes that represent BigInteger as if it was a 256-bit integer.</returns>
        public static byte[] GetBytes(BigInteger bigInteger, int byteCount = 32)
        {
            // Obtain the bytes which represent this BigInteger.
            byte[] result = bigInteger.ToByteArray();

            // We'll operate on the data in little endian (since we'll extend the array anyways and we'd have to copy the data over anyways).
            if (!BitConverter.IsLittleEndian)
                Array.Reverse(result);

            // Store the original size of the data, then resize it to the size of a word.
            int originalSize = result.Length;
            Array.Resize(ref result, byteCount);

            // BigInteger uses the most significant bit as sign and optimizes to return values like -1 as 0xFF instead of as 0xFFFF or larger (since there is no bound size, and negative values have all leading bits set)
            // Instead if we wanted to represent 256 (0xFF), we would add a leading zero byte so the sign bit comes from it, and will be zero (positive) (0x00FF), this way, BigInteger knows to represent this as a positive value.
            // Because we resized the array already, it would have added leading zero bytes which works for positive numbers, but if it's negative, all extended bits should be set, so we check for that case.

            // If the integer is negative, any extended bits should all be set.
            if (bigInteger.Sign < 0)
                for (int i = originalSize; i < result.Length; i++)
                    result[i] = 0xFF;

            // Flip the array so it is in big endian form.
            Array.Reverse(result);

            return result;
        }
    }
}

File: core0/us/sdk/wallet/java/us/gov/crypto/ec.java
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the GPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
package us.gov.crypto;
import java.util.Arrays;                                                                       // Arrays
import org.spongycastle.asn1.ASN1Integer;                                                      // ASN1Integer
import java.math.BigInteger;                                                                   // BigInteger
import org.spongycastle.util.BigIntegers;                                                      // BigIntegers
import org.spongycastle.jce.provider.BouncyCastleProvider;                                     // BouncyCastleProvider
import java.io.ByteArrayOutputStream;                                                          // ByteArrayOutputStream
import us.CFG;                                                                                 // CFG
import org.spongycastle.asn1.DERSequenceGenerator;                                             // DERSequenceGenerator
import org.spongycastle.math.ec.ECCurve;                                                       // ECCurve
import org.spongycastle.crypto.params.ECDomainParameters;                                      // ECDomainParameters
import org.spongycastle.jce.spec.ECNamedCurveParameterSpec;                                    // ECNamedCurveParameterSpec
import org.spongycastle.jce.spec.ECNamedCurveSpec;                                             // ECNamedCurveSpec
import org.spongycastle.jce.ECNamedCurveTable;                                                 // ECNamedCurveTable
import org.spongycastle.jce.spec.ECParameterSpec;                                              // ECParameterSpec
import org.spongycastle.math.ec.ECPoint;                                                       // ECPoint
import org.spongycastle.jce.ECPointUtil;                                                       // ECPointUtil
import org.spongycastle.jce.interfaces.ECPrivateKey;                                           // ECPrivateKey
import org.spongycastle.jce.spec.ECPrivateKeySpec;                                             // ECPrivateKeySpec
import org.spongycastle.jce.interfaces.ECPublicKey;                                            // ECPublicKey
import org.spongycastle.jce.spec.ECPublicKeySpec;                                              // ECPublicKeySpec
import java.security.GeneralSecurityException;                                                 // GeneralSecurityException
import org.spongycastle.util.encoders.Hex;                                                     // Hex
import static us.gov.crypto.types.*;                                                           // *
import java.security.InvalidAlgorithmParameterException;                                       // InvalidAlgorithmParameterException
import java.security.InvalidKeyException;                                                      // InvalidKeyException
import java.security.spec.InvalidKeySpecException;                                             // InvalidKeySpecException
import java.io.IOException;                                                                    // IOException
import java.security.KeyFactory;                                                               // KeyFactory
import java.security.KeyPairGenerator;                                                         // KeyPairGenerator
import java.security.KeyPair;                                                                  // KeyPair
import us.ko;                                                                                  // ko
import java.security.NoSuchAlgorithmException;                                                 // NoSuchAlgorithmException
import java.security.NoSuchProviderException;                                                  // NoSuchProviderException
import static us.ko.ok;                                                                        // ok
import java.security.PrivateKey;                                                               // PrivateKey
import java.security.Provider;                                                                 // Provider
import java.security.PublicKey;                                                                // PublicKey
import java.security.Security;                                                                 // Security
import org.spongycastle.crypto.digests.SHA256Digest;                                           // SHA256Digest
import java.security.Signature;                                                                // Signature

public class ec {
    linux_secure_random secure_random;
    KeyFactory factory;
    ECParameterSpec spec;
    ECDomainParameters curve;
    KeyPairGenerator generator;
    Signature dsa;

    public static ec instance = null;

    static void log(final String line) {            //--strip
        CFG.log_gov_crypto("ec: " + line);          //--strip
    }                                               //--strip

    public static void create_instance() {
        if (instance != null) {
            log("ec instance already created.");  //--strip
            return;
        }
        try {
            log("Initialising ec instance...");  //--strip
            Provider p = new BouncyCastleProvider();
            if (p == null) {
                System.err.println("KO 10199 BC Provider not found.");
                System.exit(1);
            }
            Security.insertProviderAt(p, 1);
            log("Added Provider");  //--strip
            instance = new ec("secp256k1", "ECDSA", "NONEwithECDSA");
        }
        catch (Exception e) {
            System.out.println("KO 10099 ECDSA: " + e.getMessage());
            System.exit(1);
        }
        log("ec instance Initialised.");  //--strip
    }

    public ec(String curve_name, String algorithm, String signature_algorithm) throws GeneralSecurityException {
        dsa = Signature.getInstance(signature_algorithm);
        spec = ECNamedCurveTable.getParameterSpec(curve_name);
        factory = KeyFactory.getInstance(algorithm);
        curve = new ECDomainParameters(spec.getCurve(), spec.getG(), spec.getN(), spec.getH());
        secure_random = new linux_secure_random();
        generator = KeyPairGenerator.getInstance(algorithm);
        generator.initialize(spec);
    }

    public BigInteger generate_private_int() {
        KeyPair keypair = generate_keypair();
        if (keypair == null) return null;
        return get_private_int(keypair);
    }

    public BigInteger get_private_int(PrivateKey k0) {
        ECPrivateKey k = (ECPrivateKey) k0;
        return k.getD();
    }

    private BigInteger get_private_int(KeyPair keypair) {
        return get_private_int(keypair.getPrivate());
    }

    public PrivateKey get_private_key(BigInteger k) {
        try {
            ECPrivateKeySpec ks = new ECPrivateKeySpec(k, spec);
            return factory.generatePrivate(ks);
        }
        catch (Exception e) {
            return null;
        }
    }

    public KeyPair generate_keypair() {
        try {
            return generator.generateKeyPair();
        }
        catch (Exception e) {
            return null;
        }
    }

    public KeyPair generate_keypair(PrivateKey k) {
        try {
            return new KeyPair(get_public_key(k), k);
        }
        catch (Exception e) {
            return null;
        }
    }

    public KeyPair get_keypair(BigInteger priv) {
        return generate_keypair(get_private_key(priv));
    }

    public byte[] generate_shared_key(PrivateKey priv, PublicKey pub, int length) {
        try {
            ECPrivateKey k = (ECPrivateKey) priv;
            ECPublicKey p = (ECPublicKey) pub;
            sha256.hash_t h = sha256.hash(multiply(p.getQ(), k.getD()).getEncoded(true));
            return Arrays.copyOf(h.value, 16);
        }
        catch (Exception e) {
            log("generate_shared_key " + e.getMessage()); //--strip
            return null;
        }
    }

    public byte[] to_vector(PublicKey pub) {
        try {
            ECPublicKey p = (ECPublicKey) pub;
            ECPoint point = p.getQ();
            return point.getEncoded(true);
        }
        catch (Exception e) {
            return null;
        }
    }

    public byte[] to_vector(PrivateKey priv) {
        try {
            ECPrivateKey k = (ECPrivateKey) priv;
            return BigIntegers.asUnsignedByteArray(32, k.getD());
        }
        catch (Exception e) {
            return null;
        }
    }

    public ripemd160.hash_t to_address(PublicKey k) {
        return ripemd160.hash(to_vector(k));
    }

    public String to_encoded_address(PublicKey k) {
        return base58.encode(to_address(k).value);
    }

    public String to_b58(PublicKey k) {
            return base58.encode(to_vector(k));
    }

    public String to_b58(PrivateKey k) {
        try {
            return base58.encode(to_vector(k));
        }
        catch (Exception e) {
            return null;
        }
    }

    public PrivateKey get_private_key(byte[] privateKey) {
        try {
            return get_private_key(BigIntegers.fromUnsignedByteArray(privateKey));
        }
        catch (Exception e) {
            return null;
        }
    }

    public PrivateKey get_private_key(byte[] buffer, int offset) {
        byte[] v = new byte[32];
        System.arraycopy(buffer, offset, v, 0, 32);
        return get_private_key(BigIntegers.fromUnsignedByteArray(v));
    }

    private ECPoint get_point(byte[] pub) {
        return spec.getCurve().decodePoint(pub);
    }

    public ECPoint multiply(ECPoint point, BigInteger d) {
        if (d.bitLength() > spec.getN().bitLength()) {
            d = d.mod(spec.getN());
        }
        return point.multiply(d);
    }

    public PublicKey get_public_key(BigInteger priv) {
        try {
            ECPoint point = multiply(spec.getG(), priv);
            return get_public_key(point);
        }
        catch (Exception e) {
            return null;
        }
    }

    public PublicKey get_public_key(PrivateKey priv) {
        try {
            BigInteger i = get_private_int(priv);
            return get_public_key(i);
        }
        catch (Exception e) {
            return null;
        }
    }

    public PublicKey get_public_key(ECPoint point) {
        try {
            ECPublicKeySpec s = new ECPublicKeySpec(point, spec);
            return factory.generatePublic(s);
        }
        catch (Exception e) {
            return null;
        }
    }

    public PublicKey get_public_key(byte[] public_key) {
        try {
            ECPoint point = get_point(public_key);
            return get_public_key(point);
        }
        catch (Exception e) {
            return null;
        }
    }

    public PublicKey get_public_key(byte[] buffer, int offset) {
        assert buffer.length >= offset + 33;
        byte[] v = new byte[33];
        System.arraycopy(buffer, offset, v, 0, 33);
        return get_public_key(v);
    }

    public boolean verify(PublicKey pub, sha256.hash_t msgh, sig_der_t sig_der) {
        log("verify msgh " + msgh.b58() + " pubk " + to_b58(pub) + " sig " + base58.encode(sig_der.value) + " siglen = " + sig_der.value.length); //--strip
        try {
            dsa.initVerify(pub);
            dsa.update(msgh.value);
            log("sig(der) length = " + sig_der.value.length + " hex " + Hex.toHexString(sig_der.value)); //--strip
            return dsa.verify(sig_der.value);
        }
        catch (Exception e) {
            log("verify failed " + e.getMessage()); //--strip
            return false;
        }
    }

    public sig_der_t sign(PrivateKey priv, sha256.hash_t msgh) {
        try {
            dsa.initSign(priv);
            dsa.update(msgh.value);
            byte[] r = dsa.sign();
            log("signature created length = " + r.length + " bytes. " + Hex.toHexString(r)); //--strip
            return new sig_der_t(r);
        }
        catch (Exception e) {
            return null;
        }
    }

}

